### ISSUE TITLE
How does tensorpack compare to the tf.estimators?

### ISSUE THREAD

**Issue by pkubik** [tensorpack/tensorpack#855]
I can see a lot of similarities between the tensorpack's trainer and its callbacks and `tf.estimators`/`tf.train` with all of its hooks. What are some benefits of using tensorpack instead of tensorflow API in case of this inversion of control? Why aren't `Callback` and `tf.train.SessionRunHook` compatible?

**Comment by ppwwyyxx** [comment]
Regarding Callbacks vs SessionRunHooks:
> Why aren't Callback and tf.train.SessionRunHook compatible?

They are compatible: you can convert between them with http://tensorpack.readthedocs.io/modules/callbacks.html#tensorpack.callbacks.HookToCallback .

Tensorpack callbacks exist mainly because it is written long before `SessionRunHook` ever exists.
It has one small extra feature about scheduling -- it has the concept of epoch which allows you to do something once a while. But you can easily implement this with SessionRunHook as well. 

Regarding trainer vs tf.estimator

I personally think `tf.Estimator` is not well designed and therefore not flexible enough. But I also think it does have something that tensorpack trainers can learn from.

Also, `tf.Estimator` does not support efficient data-parallel training, until very recently (still in contrib). However tensorpack trainer supports it from the very beginning.

**Comment by pkubik** [comment]
Thanks @ppwwyyxx . It's quite odd that they didn't simply consider integrating the tensorpack with tensorflow as they did with keras (although the keras integration is kind of broken IMHO). They didn't even mention the tensorpack in [their whitepaper](https://terrytangyuan.github.io/data/papers/tf-estimators-kdd-paper.pdf).

The biggest problem with estimators seems to be their lifecycle. It seems like they add few features, make it big on Google I/O and then forget about it for the rest of the year. E.g. I was recently thinking about implementing a run hook for printing the progress bar but it seems that there already is such thing in tensorpack.

Btw. I just found out that keras also use a similar concept of callbacks [https://www.tensorflow.org/versions/r1.8/api_docs/python/tf/keras/callbacks/Callback](https://www.tensorflow.org/versions/r1.8/api_docs/python/tf/keras/callbacks/Callback).

I guess that I'll just give the tensorpack a try.

**Comment by ppwwyyxx** [comment]
Keras is much more popular than tensorpack, so what they did is quite reasonable.

**Comment by yg320** [comment]
Hi!

I was wondering about your comment:

"I personally think `tf.Estimator` is not well designed and therefore not flexible enough. But I also think it does have something that tensorpack trainers can learn from."

Can you please elaborate about the flexibility issues on tf.Estimator that you observed? 
And also, what do you think that Tensorpack trainers can learn from tf.Estimators?
 
I'm using Tensorpack for a while now, and was considering giving tf.Estimators a try (especially because of the community)

Thanks!

**Comment by pkubik** [comment]
@yg320 It seems like the support for the estimators is going to be limited in TensorFlow 2.0. The custom estimators are discouraged in favor of the Keras models. Actually, they never appeared to have huge adoption in the community. It was always difficult to find any hints on how to use them. Note that since, TF is very popular there are a lot of low-effort tutorials which mostly paraphrase the original docs, but do not touch any details that might interest you in practice.

Source:
https://medium.com/tensorflow/whats-coming-in-tensorflow-2-0-d3663832e9b8

**Comment by ppwwyyxx** [comment]
I'm apparently not an estimator user so I may be wrong. By just looking at the documentation I can identify many flexibility issues in its design.

From https://www.tensorflow.org/api_docs/python/tf/estimator/Estimator
> model_fn: Model function. Follows the signature:
Args:
features: This is the first item returned from the input_fn passed to train, evaluate, and predict. This should be a single tf.Tensor or dict of same.
labels: This is the second item returned from the input_fn passed to train, evaluate, and predict. This should be a single tf.Tensor or dict of same (for multi-head models). If mode is tf.estimator.ModeKeys.PREDICT, labels=None will be passed. If the model_fn's signature does not accept mode, the model_fn must still be able to handle labels=None.
mode: Optional. Specifies if this training, evaluation or prediction. See tf.estimator.ModeKeys.
params: Optional dict of hyperparameters. Will receive what is passed to Estimator in params parameter. This allows to configure Estimators from hyper parameter tuning.
config: Optional estimator.RunConfig object. Will receive what is passed to Estimator as its config parameter, or a default value. Allows setting up things in your model_fn based on configuration such as num_ps_replicas, or model_dir.
Returns: tf.estimator.EstimatorSpec

Bad abstraction: it assumes all your inputs are either "features" or "labels". I often have inputs that do something else and I would feel uncomfortable calling them either "features" or "labels". Of course you can still put them into one of the two buckets but it's just ugly.
Tensorpack: everything is just "inputs"

> evaluate(
    input_fn,
    steps=None,
    hooks=None,
    checkpoint_path=None,
    name=None
)

Bad assumption: It takes an `input_fn` and will just use the existing `model_fn` with `mode=EVAL`.
So it assumes you'll want to evaluate the same `model_fn` with different `input_fn`. What if I want to evaluate also with different `model_fn` (which may still share all the trained parameters) ? Looks impossible. 
Tensorpack: support evaluation with different input + different tower function. The code release of our recent paper (https://github.com/facebookresearch/ImageNet-Adversarial-Training/) does exactly this type of evaluation.

Let's look at `EstimatorSpec` (https://www.tensorflow.org/api_docs/python/tf/estimator/EstimatorSpec) which is what `model_fn` is supposed to return:

> Depending on the value of mode, different arguments are required. Namely
For mode == ModeKeys.TRAIN: required fields are loss and train_op.
For mode == ModeKeys.EVAL: required field is loss.
For mode == ModeKeys.PREDICT: required fields are predictions.

Strong assumption: assumes that the training "runs one train_op in a loop".
Tensorpack: assumes that the training is "runs in a loop".

Bad assumption: assumes you can compute a loss in evaluation. There are counter examples in tensorpack.

> eval_metric_ops: Dict of metric results keyed by name. The values of the dict can be one of the following: (1) instance of Metric class. (2) Results of calling a metric function, namely a (metric_tensor, update_op) tuple. metric_tensor should be evaluated without any impact on state (typically is a pure computation results based on variables.). For example, it should not trigger the update_op or requires any input fetching.

Bad assumption: It assumes that metrics can be computed by ops. What if I have to compute them in Python?
Maybe there are still ways to fetch tensors in the network and compute the metrics in Python. But then it gets unclear how to get those metrics to work with the rest of estimator (e.g., send them into tensorboard).
In general TF seems to assume you want to do all the work in TF (e.g., the entire dataset module), which is often just impractical.

These are just from a brief scan of the docs. I'm not an estimator user and these opinions may be subjective or wrong. 
And I would definitely agree if you say that estimator is good enough for 95% of use cases, or 100% good enough for your use cases. I think Keras is also 100% good enough for perhaps 90% of TF users.
